File: Src/WitsmlExplorer.Frontend/models/componentType.ts
import DataSourceConfigurationSet from "models/dataWorkOrder/dataSourceConfigurationSet";
import Fluid from "models/fluid";
import GeologyInterval from "models/geologyInterval";
import LogCurveInfo from "models/logCurveInfo";
import { ObjectType } from "models/objectType";
import TrajectoryStation from "models/trajectoryStation";
import TubularComponent from "models/tubularComponent";
import WbGeometrySection from "models/wbGeometrySection";

export enum ComponentType {
  GeologyInterval = "GeologyInterval",
  Mnemonic = "Mnemonic",
  TrajectoryStation = "TrajectoryStation",
  TubularComponent = "TubularComponent",
  WbGeometrySection = "WbGeometrySection",
  Fluid = "Fluid",
  DataSourceConfigurationSet = "DataSourceConfigurationSet"
}

export const getParentType = (componentType: ComponentType): ObjectType => {
  switch (componentType) {
    case ComponentType.GeologyInterval:
      return ObjectType.MudLog;
    case ComponentType.Mnemonic:
      return ObjectType.Log;
    case ComponentType.TrajectoryStation:
      return ObjectType.Trajectory;
    case ComponentType.TubularComponent:
      return ObjectType.Tubular;
    case ComponentType.WbGeometrySection:
      return ObjectType.WbGeometry;
    case ComponentType.Fluid:
      return ObjectType.FluidsReport;
    case ComponentType.DataSourceConfigurationSet:
      return ObjectType.DataWorkOrder;
    default:
      return undefined;
  }
};

export type ComponentTypeToModel = {
  [ComponentType.GeologyInterval]: GeologyInterval;
  [ComponentType.Mnemonic]: LogCurveInfo;
  [ComponentType.TrajectoryStation]: TrajectoryStation;
  [ComponentType.TubularComponent]: TubularComponent;
  [ComponentType.WbGeometrySection]: WbGeometrySection;
  [ComponentType.Fluid]: Fluid;
  [ComponentType.DataSourceConfigurationSet]: DataSourceConfigurationSet;
};

File: Src/WitsmlExplorer.Api/Models/ComponentType.cs
using System;
using System.Linq;
using System.Text.Json.Serialization;

namespace WitsmlExplorer.Api.Models
{
    [JsonConverter(typeof(JsonStringEnumConverter))]
    public enum ComponentType
    {
        GeologyInterval,
        Mnemonic,
        TrajectoryStation,
        TubularComponent,
        WbGeometrySection,
        Fluid
    }

    public static class ComponentTypeExtensions
    {
        public static string ToPluralLowercase(this ComponentType componentType)
        {
            string lower = componentType.ToString().ToLowerInvariant();
            return lower.Last() == 'y' ? lower.Remove(lower.Length - 1) + "ies" : lower + "s";
        }

        public static EntityType ToParentType(this ComponentType componentType)
        {
            return componentType switch
            {
                ComponentType.GeologyInterval => EntityType.MudLog,
                ComponentType.Mnemonic => EntityType.Log,
                ComponentType.TrajectoryStation => EntityType.Trajectory,
                ComponentType.TubularComponent => EntityType.Tubular,
                ComponentType.WbGeometrySection => EntityType.WbGeometry,
                ComponentType.Fluid => EntityType.FluidsReport,
                _ => throw new ArgumentException($"Invalid component type {componentType}"),
            };
        }
    }
}
